### PR TITLE
LF-4542: Allow custom task to have related animal ids on the backend

### PR DESCRIPTION
**Pull request by SayakaOno**
**Description**

The major changes are to:
- apply the same checks for animal tasks to custom tasks in POST and PATCH `/patch_due_date`.
- enable custom tasks completion while removing all animals/batches from the task.

============

- GET
   - add tests
- POST
   - Add validation to avoid `animals` and `animal_batches` to be sent from the client (this should be moved to the middleware in the future). Without this, animals/batches can be added to any tasks. (this will also be run for the completion API)
- PATCH
  - complete
     - Move the validations for missing animals/batches and invalid `complete_date` to the middleware (`checkAnimalCompleteTask`).
     - Update `upsertGraph` options to relate/unrelate `animals` and `animal_batches`. (This shouldn't affect other task types that won't have animals or batches)
     - Update `formatAnimalAndBatchIds` to unrelate animals/batches when `null` is sent for `related_animal_ids` and `related_batch_ids`
  - due_date
    - Simplify `checkDueDate` function to handle custom tasks. (the check will be run for all tasks that have animals or batches)

Jira link: https://lite-farm.atlassian.net/browse/LF-4542

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [x] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/api/src/controllers/taskController.js [hunk 1]
```diff
@@ -152,27 +153,10 @@ async function updateTaskWithCompletedData(
...
-        data.complete_date,
-        (data.animals || []).map(({ id }) => id),
-        (data.animal_batches || []).map(({ id }) => id),
-      );
-
-      if (!isValidDate) {
-        throw customError(
-          `complete_date must be on or after the animals' birth and brought-in dates`,
-        );
-      }
```
Comment (SayakaOno) [inline packages/api/src/controllers/taskController.js:174]: These checks were moved to the middleware.

File: packages/api/src/controllers/taskController.js [hunk 2]
```diff
@@ -520,11 +509,15 @@ const taskController = {
   },
 
   formatAnimalAndBatchIds(data) {
-    if (data.related_animal_ids) {
-      data.animals = [...new Set(data.related_animal_ids)].map((id) => ({ id }));
+    if ('related_animal_ids' in data) {
+      data.animals = data.related_animal_ids
+        ? [...new Set(data.related_animal_ids)].map((id) => ({ id }))
+        : null;
```
Comment (SayakaOno) [inline packages/api/src/controllers/taskController.js:515]: If `related_animal_ids` is null (the intention is to remove the relation), set `data.animals` to null. The relation will be removed by sending `null` to `upsertGraph`.

File: packages/api/src/controllers/taskController.js [hunk 3]
```diff
@@ -534,6 +527,11 @@ const taskController = {
   },
 
   async checkCustomDependencies(typeOfTask, data, farm_id) {
+    // TODO: Move this validation to checkCreateTask and checkCompleteTask and apply the middleware to all relevant routes.
+    if ('animals' in data || 'animal_batches' in data) {
+      throw customError(`Invalid field: "animals" or "animal_batches" should not be included.`);
+    }
```
Comment (SayakaOno) [inline packages/api/src/controllers/taskController.js:533]: You would be able to add animals or batches to any types of tasks by including `animals: [{id: xxx}]` in the API request without this check.

File: packages/api/tests/animalTask.test.js [hunk 4]
```diff
@@ -602,11 +626,31 @@ describe('Animal task tests', () => {
...
+          const animals = await knex('animal')
+            .select('id')
+            .whereIn('id', beforeAnimalIds)
+            .andWhereNot('deleted', true);
+          expect(animals.length).toBe(beforeAnimalIds.length);
+          const batches = await knex('animal_batch')
+            .select('id')
+            .whereIn('id', beforeBatchIds)
+            .andWhereNot('deleted', true);
+          expect(batches.length).toBe(beforeBatchIds.length);
```
Comment (SayakaOno) [inline packages/api/tests/animalTask.test.js:653]: Made the checks more strict by checking the data in the DB.

File: packages/api/tests/task.test.js [hunk 5]
```diff
@@ -1949,11 +1949,13 @@ describe('Task tests', () => {
           }),
         );
 
+        const { animals, animal_batches, ...patchTaskData } = createdTask;
```
Comment (SayakaOno) [inline packages/api/tests/task.test.js:1952]: `createdTask` (Post request's response body) now includes animals and animal_batches that shouldn't be included in the patch request. This is the adjustment.

**Review by antsgar — APPROVED** [review]
Tested it together with the UI changes and it worked perfectly, great job!

**Review by kathyavini — APPROVED** [review]
It works perfectly with Duncan's frontend change and with the missing pieces of the custom flow ❤️
